### ISSUE TITLE
BLAS and LAPACK compatibility with main tree

### ISSUE THREAD

**Issue by TheChymera** [gentoo/sci#805]
Allegedly [[1]]:

> The BLAS and LAPACK stack in the science overlay are incompatible with the BLAS and LAPACK stack from the main tree.

I have confrimed this in the past, but I am currently looking to set up a machine from scratch (and maybe to produce an auto-script or documented step-by-step manual install workflow for scientific Gentoo) - and I'm wondering if this is really still necessary.
If so, could we do anything about it? 

[1]: https://wiki.gentoo.org/wiki/User_talk:Houseofsuns

**Comment by kiwifb** [comment]
Not only is this necessary but I think we need better documentation than that link. Something more visible. I have been working on and off on trying to get something together in the style of the kde overlay where you can just link to some files from the overlay in `/etc/portage/package.*`.

And of course that could be scripted once you add the overlay.

**Comment by kiwifb** [comment]
See https://github.com/kiwifb/sci/tree/blas-doc for some work in progress - look under `Documentation`. Note that since then I have worked on `eigen` (awaiting for merge) which negates the need for separate files. But it need to come with instructions like the one in [https://wiki.gentoo.org/wiki/User_talk:Houseofsuns](url)

**Comment by heroxbd** [comment]
No, they are not compatible.  We had better prioritize our efforts to merge the framework back to gentoo main repo.

**Comment by kiwifb** [comment]
What needs to happen for that?

**Comment by heroxbd** [comment]
The most important and out of the control of science team part is to get eselect extensions upstreamed.

**Comment by lkraav** [comment]
@heroxbd what's the "out of control" part about, is there any gentoo bugzilla issue to review and get to know the situation?

**Comment by heroxbd** [comment]
@lkraav Here is the snapshot of situation.  We have some discussions how to move forward on the IRC.  The consensus now is to adopt the `multibuild.eclass` similar to `PYTHON_TARGETS` to BLAS and LAPACK.

**Comment by kiwifb** [comment]
I have now been thinking about this for a couple of hours. Is there a summary or some sort of document yet? Right now I am feeling like "shoot me now". All I can think is that it will add a big layer of complexity unless you only ever allow one target. But then `multibuild` wouldn't be any use.

Speaking of `python` I have python packages that depend on blas/lapack, all I can imagine is a big mess where `PYTHONPATH` will probably have to be augmented for the appropriate blas/lapack.

But I should wait for some specs before going insane.

**Comment by heroxbd** [comment]
There is no documentation, we should start to write one.  

> All I can think is that it will add a big layer of complexity unless you only ever allow one target. But then multibuild wouldn't be any use.

Not necessarily to have multiple implementations, think of PYTHON_SINGLE_TARGET.

The design flaw of alternative.eclass is being runtime.  For example, when we want to build `numpy` with `OpenBLAS` but `R` with  `refblas`, we have to
  1. eselect blas set <openblas>
  2. emerge numpy
  3. eselect blas set <refblas>
  4. emerge R
No way to write it down in a configuration file before just an `emerge numpy R`.

This design flaw comes from the fact that, alternative.eclass was originally designed to let different blas implementations interchangable at runtime.  But when the ABI compatibility issues arise, it evolved into a compile time selection.

**Comment by kiwifb** [comment]
Yes the current system has flaws. And I see why you want a multibuild like system now. So long as it preserve the use of pkgconfig in the first place, I guess we'll be able to get stuff done.

Will we be able to comment or offer criticism on the reference document?

**Comment by heroxbd** [comment]
Yes, sure.  That will be the best way to move forward.

**Comment by sfabbro** [comment]
Initially the idea with alternative/eselect was to also to dynamically create a linker script with at selection time. It was never implemented. Reading very briefly at the ld manual, I am not sure anymore. Anyone knows if we can actually force a SONAME to a linker script? 

If linker script idea does not work, one could also create an on-the-fly library stub at selection time. For example, an eselect blas set openblas_threads would actually do  under the hood something like:

gcc -shared  -lopenblas_threads -Wl,--soname=libblas.so.3 -o libblas.so.3

That could also be a solution for runtime issues without massive ebuild changes.

**Comment by heroxbd** [comment]
@sfabbro Hi Sébastien, thank you for clarifying this.  So, is the "ABI problem" (I have learned from you before) just the inconsistency of SONAME?

**Comment by Alessandro-Barbieri** [comment]
Why not use flexiblas?

**Comment by kiwifb** [comment]
Interesting, slightly different in scope than what has been talked so far. It would need some tweaks to do what we want I think, but interesting.

**Comment by heroxbd** [comment]
See also https://bugs.gentoo.org/632624

**Comment by Doeme** [comment]
Hi.
@heroxbd mentioned the consensus being doing something like PYTHON_TARGETS. 
Does this imply building (or at least linking) each program/library seperately for every blas*lapack-version enabled in USE, or rather deciding on one "best" implementation and link the program (and all its dependencies) against that single one?

The former case poses the question on the file-hierarchy being used (something like python?), and if this is even reasonable considering the cartesian product Python⨯BLAS⨯LAPACK blowing up.
Also, "import numpy-openblas-reflapack"" seems rather silly, so maybe this case would be most reasonably handled with the flexiblas-approach.

The latter might

**Comment by kiwifb** [comment]
@heroxbd also mentioned `PYTHON_SINGLE_TARGET`. In fact that's the only real sane option in my opinion. But flexiblas could do some of that too.

Also something that I don't really want to lose is the use of .pc files for building. Some of the stuff needed is already there. Every blas/lapack has a different library name so they can be co-installed. The eclass and `{BLAS,CBLAS,LAPACK,...}_TARGET}` variable (I am getting ahead of myself there) could just point to the right .pc files for building.

**Comment by Doeme** [comment]
How should the hierarchy look then for programs/libraries linking against LAPACK? /usr/lib/libfoo-reflapack-openblas.so?
What about e.g. numpy with "pythonm.y/numpy/linalg/_umath_linalg.so"? Make a "pythonm.n/numpy-reflapack-openblas/numpy/linalg/_umath_linalg.so"?
Should we then do eselects where one can choose a default (BLAS,LAPACK)-pair for all applications?

**Comment by kiwifb** [comment]
Scenarios as I see them:

1. flexiblas, then blas,lapack can be changed at runtime anywhere by an eselect switch so long as the matching blas/lapack is installed.
2. Not using flexiblas. May not need eselect. Set `{BLAS,LAPACK}_TARGET` in `make.conf` with some default if not defined. Can be overridden package per package through env/package.env like `PYTHON_SINGLE_TARGET`.

In scenario 2 a package is compiled against a given blas/lapack. If you want to change it, you need to rebuild.

**Comment by Doeme** [comment]
But 2. would then enforce a single global BLAS/LAPACK-implementation on the whole system?
Just changing it via package.env would most likely result in runtime-errors while resolving symbols when the dependencies have been built with other implementations.
Now that I think about it: Is it just coincidence (or rather: luckily disjoint dependency trees) that allowed things like described in https://github.com/gentoo/sci/issues/805#issuecomment-332758079 ?

**Comment by kiwifb** [comment]
Disjoint dependency trees. We always have given the opportunity to the end user to shoot themselves in the foot.

**Comment by Doeme** [comment]
Ok, cool. So IMHO the only option to go was my or your second one, because it automatically implies the first one (the user could, if he or she so wished, build against flexiblas).

**Comment by sfabbro** [comment]
I would recommend to check whether the problem is a soname issue only. If it is, here is the third scenario:

  3.  Implement dynamic soname with eselect and alternatives. If that works, it would apply more than only to blas/lapack such as flexiblas, it would fix the joint dependency problem in #805, it would require only minimal changes of ebuilds (i.e. no need of BLAS_TARGET on all reverse deps), and there would be no need of rebuilding everything on a change of blas. The .pc files could be kept, but they would only point to /usr/lib/libblas.so which would be basically a re-sonamed of /usr/lib/libopenblas-threads.so. That would make packaging new apps easier, since /usr/lib/libblas.so would always exist.

**Comment by heroxbd** [comment]
@sfabbro How about coerce all the BLAS implementations into the same soname as `libblas.so.0`?

**Comment by kiwifb** [comment]
How do you do that with proprietary library like intel MKL? I can only think of a shim which bring us back to stuff like flexiblas.

**Comment by heroxbd** [comment]
@Doeme 

> Also, "import numpy-openblas-reflapack"" seems rather silly, so maybe this case would be most reasonably handled with the flexiblas-approach.

If you could interpret import numpy as numpy compiled against `BLAS_SINGLE_TARGET` and `LAPACK_SINGLE_TARGET`, that will be less of a problem.

Flexblas uses `dlopen`, making it an independent angle to attack this problem IMHO.

**Comment by Doeme** [comment]
@sfabbro This - I think - would again fold back onto solution 2., since we would best make a new blas-package that provides some sort of common header, which then were one of the options in BLAS_IMPLS. This would then ensure that we are not e.g. compiling a library against openblas and eselect something other than openblas for a program using that library, resulting again in symbol lookup errors. Or even worse: Compile library A against openblas, B against refblas and program C against the eselectable variant, which would always lead to symbol lookup errors.

@heroxbd So would you go for the cartesian product variant, building every library for all selected (BLAS,LAPACK,Python)-tuples?

By the way: I guess there are/could be packages only compatible with a single BLAS implementation?

**Comment by Doeme** [comment]
As can be seen in those strange timeline-thingies, I tried out implementing a blas-single-implementation eclass. The results can be diffed here:
https://github.com/gentoo/sci/compare/master...Doeme:master

This requires the ::science versions of the packets being installed, i.e. the proper .pc-files in /usr/lib/pkgconfig/

As you can see this is just a draft, therefore comments aren't up to par and, in fact, non-existent :)
But your comments on the design will be highly appreciated.

**Comment by TheChymera** [comment]
Hi guys @kiwifb @heroxbd @sfabbro we would be looking to get this off the ground in order to ease system construction for scientific gentoo. This looks really good to me, but we'd be happy for your feedback (and help in getting this - or a more advanced solution based on this - pulled).

**Comment by kiwifb** [comment]
I am a bit busy, but I can make some time. It would be good to have a chat at some point on IRC

**Comment by kiwifb** [comment]
I cannot comment directly on the commits. First impression is Ok but `REQUIRED_USE="$BLAS_REQ_USE"` shouldn't need to be in the ebuild, we can probably add it from inside the eclass. It may get tricky if `blas` is optional.
Same with `BLAS_DEP`, it should only be needed in some cases like for the similar case with python.

**Comment by Alessandro-Barbieri** [comment]
My 2 cent:

don't install every `BLAS` implementation supported by every package but only the best, where best is chosen by the user in a way similar to python: with a file containing a list of priorities

example for `BLAS`

```
openblas
atlas
refblas
```

then all packages will be built with `openBLAS` if they support it, if not with `ATLAS`, if not with `BLAS`

**Comment by kiwifb** [comment]
We have two possible designs
* The programmatic way, which I believe started it all. Package will build against one specific blas implementation specified by a `BLAS` variable. This can be set in `make.conf` or on the command line or in `env`. The eclass will set a sensible default if `BLAS` is not set, and take care of the setup for a given value of `BLAS`. This only need an eclass. It may need a way to specify a particular value of `BLAS` inside an ebuild in the rare case you need a specific one (it shouldn't unless you rely on non-blas part shipped with the library - like mkl).

* The runtime switching way. We have an eclass and a `eselect` module. Each blas register with eselect when built. eselect set a default runtime value for blas. Package are built against a shim that just expose a selected blas. Bonus point if eselected default can be overridden from user space by an environment variable. 

Does that sound like a fair description of our options? Implementation details could vary a bit.

**Comment by Doeme** [comment]
For completeness sake: There still is the third option of building every package n-times for every combination of BLAS/Lapack/Python with a proper namespace for the files.

Also, as mentioned before: Point 1 and 2 are not mutually exclusive. Implementing Point 1 with additional runtime-switchable blas-implementation will yield Point 2 (given that all desired packages are compatible with that additional implementation). This method has the "advantage" of properly handling packages that are only compatible with one specific BLAS-Implementation

**Comment by Doeme** [comment]
Oh, and @kiwifb: the behaviour currently implemented in the eclass is afaik the exact same as in the python eclasses (i.e. neither the dependencies nor the required_use get added automatically).
One goal I had was to do it in the same way as the python eclasses, as people are probably used to it by now.

**Comment by Doeme** [comment]
In my fork I converted all packages in the science overlay to my proposed eclass:
https://github.com/gentoo/sci/compare/master...Doeme:master

Things to do now:
1) Properly test it,  which depends on 
2) Convert the main portage tree to eclass
3) Think on what to do about scalapack
4) ????
5) Non-Profit

**Comment by TheChymera** [comment]
> 1. Properly test it, which depends on

Which depends on what? I assume you don't mean the second bullet point, since then it would be first, not second in the resolution order.

> 3. Wouldn't the proper time to think about it be before we commit to any one solution model?

**Comment by Doeme** [comment]
Yes, this is exactly what I meant. Sorry If the ordering is confusing. For your peace of mind I changed the list-type to an unordered one.

The problem with scalapack was: I didn't know about it until the very end of the conversion.
But it occurs only very rarely and should be easily (and without breakages) appendable to the current solution.

**Comment by heroxbd** [comment]
> And what would be the proper approach for the CBLAS or CLapack(Lapacke) problem?
> Add a variable "BLAS_USE_CBLAS" that, if set, adds dependencies to the c-headers of the BLAS-implementation (i.e. blas_refblas? ( sci-libs/blas-reference ) -> blas_refblas? ( sci-libs/blas-reference sci-libs/cblas-reference )) or should we rather add a cblas.eclass?

Is there any use case to use BLAS fortran API of OpenBLAS and CBLAS C API of ATLAS?  If not, shall we impose that all the BLAS_TARGET and CBLAS_TARGET should always be the same?  (or BLAS_USE_CBLAS===1)

**Comment by kiwifb** [comment]
While that feels a bit silly, you can do something like that right now. The thing is, a program or a library will use either the fortran API or the C API but not both at the same time. There is no point in that.

**Comment by Doeme** [comment]
Don't all C-APIs depend on an underlying Fortran API? We then would just have to say to the implementation: "Hey, I use C, install the .h's please".

(by the way, how is the Fortran API defined? I don't see any .f[0-9][0-9]-files installed by the packages. Do programs just link to the library at random? Or do they provide their own function headers)

By the way: I'm currently writing a GLEP, If you want to look it over: https://github.com/Doeme/NeuroGentooProject/blob/master/BLAS_Lapack/glep-%254d.rst (WIP)

**Comment by kiwifb** [comment]
@Doeme historically the FORTRAN API came first. And FORTRAN doesn't really have a concept of headers.

But not all implementation depend on an underlying fortran implementation. ATLAS is actually a C library with a FORTRAN layer on top. `gslcblas` is just a C library without a FORTRAN counterpart. On the other hand cblas-reference is based on the FORTRAN code.

**Comment by heroxbd** [comment]
Then I think we can treat the C API and fortran API as a single abstract compatibility unit.

`BLAS_COMPAT=*fortran` means compatible with any BLAS that provides a fortran API.
`BLAS_COMPAT=*c` means compatible with any BLAS that provides a C API.
In special cases `BLAS_COMPAT="openblas atlas"`.

**Comment by kiwifb** [comment]
I haven't looked at the GLEP yet but could we make `BLAS_COMPAT` more like `MULTILIB_COMPAT` than `PYTHON_COMPAT`?

PYTHON_COMPAT explicitly list all compatible implementation. MULTILIB_COMPAT on the other hand, from the eclass man page "List of multilib ABIs supported by the ebuild. If unset, defaults to all ABIs supported by the eclass." So by default every implementation is supported unless you put a limit.

Let's not forget that we go down a path that is not actually multibuild, unlike python and multilib.

**Comment by Doeme** [comment]
@kiwifb This is actually really reasonable, since most packages are compatible with all implementations

@heroxbd And for the case where we want the C API of a specific implementation(s)? I would have suggested looking whether CBLAS_COMPAT or BLAS_COMPAT is defined, but this doesn't work with kiwifbs suggestion anymore.

**Comment by heroxbd** [comment]
> And for the case where we want the C API of a specific implementation(s)? I would have suggested looking whether CBLAS_COMPAT or BLAS_COMPAT is defined, but this doesn't work with kiwifbs suggestion anymore.

We can have `BLAS_COMPAT=*` mean "I am compatible with whatever API, I will pick fortran or c or any future language when any is available".  And let `blas.eclass` define `BLAS_COMPAT=*` by default.

**Comment by Doeme** [comment]
And requesting the c-version of e.g. refblas? BLAS_COMPAT='c:refblas'? Then maybe use BLAS_COMPAT='c:*' for any of the c-headers.

The most frequent case is BLAS_COMPAT="f:*", maybe we should rather add that as the default?

**Comment by heroxbd** [comment]
> And requesting the c-version of e.g. refblas? BLAS_COMPAT='c:refblas'? Then maybe use BLAS_COMPAT='c:*' for any of the c-headers.

The c-version of `refblas` is `cblas-reference` (or whatever), just write `BLAS_COMPAT=cblas-reference`.  No need to explicitly spell out c: or f: when referring to a single implementation.  `c:` or `f:` is only useful for `c:*` or `f:*`.

> The most frequent case is BLAS_COMPAT="f:*", maybe we should rather add that as the default?

I have thought of that, but what if 10 years from now, no one uses fortran anymore, and we want to change the default to `BLAS_COMPAT` to `c:*`, a lot of the ebuilds should be modified.... Well on a second thought, never mind, fortran will still be there 10 years from now and we don't have that many ebuilds to modify.  So great, let's default to `f:*`.

**Comment by Doeme** [comment]
> I have thought of that, but what if 10 years from now, no one uses fortran anymore, and we want to change the default to BLAS_COMPAT to c:*, a lot of the ebuilds should be modified....

This _should_ be easily scriptable.

>The c-version of refblas is cblas-reference (or whatever), just write BLAS_COMPAT=cblas-reference.
What if the package is compatible with atlas and foobarmagicblas and needs the c-headers?
Whilst most implementations will install the c-headers automatically, I kind of feel its lacking when the ebuild can't specify that.

**Comment by heroxbd** [comment]
Hey guys, I have examined the [debian approach](https://wiki.debian.org/DebianScience/LinearAlgebraLibraries).  As a binary distribution, it is providing runtime switches of different BLAS implementations, which proves @sfabbro 's speculation that the problem is a soname issue only.  In Debian,

```
/usr/lib/libblas.so.3 -> /etc/alternatives/libblas.so.3
/etc/alternatives/libblas.so.3 -> /usr/lib/libblas/libblas.so.3
```
Python `numpy.linalg` which is linked against `libblas.so.3` works.
```
# update-alternatives --config libblas.so.3
There are 2 choices for the alternative libblas.so.3 (providing /usr/lib/libblas.so.3).

  Selection    Path                                 Priority   Status
------------------------------------------------------------
  0            /usr/lib/openblas-base/libblas.so.3   40        auto mode
* 1            /usr/lib/libblas/libblas.so.3         10        manual mode
  2            /usr/lib/openblas-base/libblas.so.3   40        manual mode
```
After selecting 0:

`/etc/alternatives/libblas.so.3 -> /usr/lib/openblas-base/libblas.so.3`

Python `numpy.linalg` still works. And from lsof, it opened /usr/lib/openblas-base/libblas.so.3.

Quoting @sfabbro 's comments above:
> I would recommend to check whether the problem is a soname issue only. If it is, here is the third scenario:
> 
>   3.  Implement dynamic soname with eselect and alternatives. If that works, it would apply more than only to blas/lapack such as flexiblas, it would fix the joint dependency problem in #805, it would require only minimal changes of ebuilds (i.e. no need of BLAS_TARGET on all reverse deps), and there would be no need of rebuilding everything on a change of blas. The .pc files could be kept, but they would only point to /usr/lib/libblas.so which would be basically a re-sonamed of /usr/lib/libopenblas-threads.so. That would make packaging new apps easier, since /usr/lib/libblas.so would always exist.

**Comment by Doeme** [comment]
@heroxbd this I am aware of, but it has drawbacks: It doesn't allow a package to specify compatibilities and it may not get mainlined: https://bugs.gentoo.org/632624

We can still apply dynamic soname-stuff by introducing a new blas-implementation (e.g. blas-soname-switch) that allows this, as well as the flexiblas thingie.

**Comment by heroxbd** [comment]
> @heroxbd this I am aware of, but it has drawbacks: It doesn't allow a package to specify compatibilities and it may not get mainlined: https://bugs.gentoo.org/632624

The QA team is against manipulating files in `/usr` at runtime, the Debian approach modifies links in `/etc/alternative`.  There is no conflict.

Before @sfabbro raised this, I have got the feeling that BLAS are only API and not ABI compatible from a `#gentoo-science` IRC meeting.  But it seems that I was wrong.

> We can still apply dynamic soname-stuff by introducing a new blas-implementation (e.g. blas-soname-switch) that allows this, as well as the flexiblas thingie.

I am confused.  If every implementation of BLAS are ABI compatible up to a soname, what does flexiblas achieve?
 
> By the way, the selective compatibility happens:
> 
> - https://github.com/gentoo/sci/blob/master/sci-chemistry/modelfree/modelfree-4.20.ebuild
> - https://github.com/gentoo/sci/blob/master/sci-astronomy/psfex/psfex-9999.ebuild
> - https://github.com/gentoo/sci/blob/master/sci-chemistry/cpmgfit/cpmgfit-1.43.ebuild
> - https://github.com/gentoo/sci/blob/master/sci-chemistry/curvefit/curvefit-1.40.ebuild

Ha, that is an interesting hack.

**Comment by Alessandro-Barbieri** [comment]
> Now, we need a real list of packages that are not compatible with all the BLAS implementations. From the Debian survey, I suspect that such a package does not exist.

https://github.com/gentoo/sci/issues/20

**Comment by Doeme** [comment]
```
dominik@X200 /tmp $ objdump -T /usr/lib/libopenblas_threads.so | grep openblas
/usr/lib/libopenblas_threads.so:     file format elf64-x86-64
0000000000272960 g    DF .text	0000000000000043  Base        openblas_set_num_threads_
0000000000273b50 g    DF .text	0000000000000041  Base        openblas_set_num_threads
0000000000272cb0 g    DF .text	0000000000000043  Base        openblas_block_factor
0000000000272a00 g    DF .text	0000000000000041  Base        openblas_get_num_procs_
0000000000272d00 g    DF .text	0000000000000043  Base        openblas_num_threads_env
0000000000272b40 g    DF .text	0000000000000042  Base        openblas_get_parallel_
0000000000272af0 g    DF .text	0000000000000042  Base        openblas_get_parallel
0000000000272a50 g    DF .text	0000000000000044  Base        openblas_get_config
0000000000272c60 g    DF .text	0000000000000043  Base        openblas_thread_timeout
0000000000272d50 g    DF .text	0000000000000043  Base        openblas_goto_num_threads_env
0000000000272b90 g    DF .text	000000000000007d  Base        openblas_warning
0000000000272df0 g    DF .text	000000000000014d  Base        openblas_read_env
0000000000272220 g    DF .text	0000000000000041  Base        openblas_get_num_procs
0000000000272c10 g    DF .text	0000000000000043  Base        openblas_verbose
0000000000272aa0 g    DF .text	0000000000000044  Base        openblas_get_corename
0000000000272da0 g    DF .text	0000000000000043  Base        openblas_omp_num_threads_env
0000000000272270 g    DF .text	000000000000004b  Base        openblas_get_num_threads
00000000002729b0 g    DF .text	0000000000000041  Base        openblas_get_num_threads_
00000000002720a0 g    DF .text	0000000000000077  Base        openblas_fork_handler
```

If a program uses these symbols, they might not get resolved.

**Comment by heroxbd** [comment]
Okay, upon deeper investigation, it looks like we cannot get shared-library ABI level interchangeability for free.  OpenBLAS, for example, needs to be patched:

https://sources.debian.net/src/openblas/0.2.20%2Bds-2/debian/patches/shared-blas-lapack.patch/

**Comment by Doeme** [comment]
What they propose is the following:
* We have multiple providers, whose libraries that get stripped down to a common symbol-table (i.e. the reference by netlib).
* For packages which need the specialized symbols we retain the original library and link against those
* If we link to both the specialized and the generic version, we let the dynamic linker decide on what symbols to actually use for the BLAS-routines.

Anyway, even if we do it like the debian folk, we need a way to specify against which libraries the packages link (Do they link against the common one or against the specialized one), which is basically the BLAS_COMPAT variable, just without propagating the use-flag down the dependency graph.

**Comment by kiwifb** [comment]
Going back for a moment to compatibility problems. Out of the 4 packages pointed by @Doeme 3 are from sci-chemistry that are actually binary packages. They even rely on gcc-4.1 which is masked.
The last one I think could work with more generic lapack, but would have required significant work. Probably at the configure script level.

@Alessandro-Barbieri the problem you mention should actually be tackled as well. The problem is with the int64 variant. Most packages, including R, will want int32. Some can or require int64 (julia comes to mind). We may need a `BLAS_REQUIRES`, like there is a `PYTHON_REQUIRE` to signal the need for a particular variant. Possible requires could be int64, openmp, threads and to link it to some other part of the discussion `c` or `fortran` :) And last of all cuda. FORTRAN may not go away very fast but some form of blas over GPU will almost certainly become widespread.

Which brings to the other point of the stack. There are 4 main components:
* blas (fortran)
* cblas (c)
* lapack (fortran)
* lapacke (c)
plus scalapack (mpi-lapack style functionality).
Left to its own devices openblas will build all 4 (inside the single library too). I think MKL also have all four plus scalapack elements in separate libraries. ATLAS only has the first 3 and provides an "alternative" to lapacke usually refereed to as clapack (thank goodness scipy doesn't look for it anymore).
I guess I am getting at the notion of whether BLAS_COMPAT is pointing to a whole stack or a ranked list of libraries for every components.

For the dynamic soname possibility, can it be made to work with binary implementations like MKL or other proprietary stuff.

**Comment by Doeme** [comment]
>Going back for a moment to compatibility problems. Out of the 4 packages pointed by @Doeme, 3 are from sci-chemistry that are actually binary packages. They even rely on gcc-4.1 which is masked.
 The last one I think could work with more generic lapack, but would have required significant work. Probably at the configure script level. is using ATLAS' clapack, bother.

There is also:
https://github.com/gentoo/sci/blob/master/sci-misc/kaldi/kaldi-5.1.65.ebuild#L41
No Idea why I missed that in the first run, it's actually the most interesting.
But in the main tree I guess most packages are compatible with the reference versions (some have mkl USE flags), hence with any implementation.

> I guess I am getting at the notion of whether BLAS_COMPAT is pointing to a whole stack or a ranked list of libraries for every components.

So If I understand you correctly, you're weighing whether to do:
- `BLAS_LAPACK_COMPAT="reference open(blas) atlas..."`  Here, we would still have to somehow say that a package needs blas, cblas, lapack, lapacke or clapack(maybe packages using the last could just depend on atlas and be done with it), since the reference-stack is split up into multiple packages, and we probably do not want to pull the whole stack into the dependency tree only because a package needs blas
- or `CBLAS_COMPAT="openblas reference"` `LAPACKE_COMPAT="mkl reference"`

>For the dynamic soname possibility, can it be made to work with binary implementations like MKL or other proprietary stuff.

I don't see why it shouldn't.

**Comment by kiwifb** [comment]
Hum, `kaldi` is indeed an interesting case. I would to look at it more closely to see if it could be "normalized". 

I guess I had to express some thoughts about stacks but I am not sure how useful it is. Bear with me since I am thinking aloud and trying to articulate something as I go along. 

It is more about building and installing the implementations in a way than using them. The current atlas ebuild can provide 3 out of 4 elements of the stack. openblas can provide 2, but with a tweak could provide 4 in one go, rather than being associated with two reference packages.
Stuff like lapack-reference is becoming a consumer of BLAS_COMPAT itself if it stays separate.

So you could have stacks where you build/install all the 4 elements based on one base implementation. MKL, atlas (+ lapacke build against atlas), openblas, reference. And then there are some oddballs like gslcblas (cblas) and eigen (blas). You lose the possibility of mixing and matching but you may gain in consistency. It is not uncommon to link to both blas and lapack (as in using symbols of both not just linking to blas because lapack depends on it) for example. Currently you can do something like use openblas for blas and then atlas for lapack. I am not sure how long your sanity can last doing stuff like that. 

On the other hand, if you implement some runtime switching, it may not matter much if the stack elements are kept separate - apart from MKL and possibly atlas's lapack. 
Apart from MKL, lapack is usually just lapack-reference compiled against a base blas implementation (with a tweak for atlas). In the runtime switching scenario, lapack-reference will just always be linked to the currently selected blas anyway.

**Comment by heroxbd** [comment]
@Doeme, regarding the Debian runtime swtich approach, I think you have the `BLAS_TARGETS` approach defended.

@kiwifb I am not sure if we go as far as gluing `blas` and `lapack`. Unlike the `c` API and `fortran` API, `blas` and `lapack` are loosely coupled together.  I haven't analyzed the rationales behind MKL shiping both `blas` and `lapack`, and `OpenBLAS` providing part of `lapack` functions. I suspect it to be a mere bundle for dependency convenience.  The development of `blis` and `libframe` on the other hand, has a clean separation between `blas` and `lapack`.  Besides, `sci-libs/scikits_learn` links only to `blas`.

@Doeme: regarding `BLAS_USEDEP`, do we really need it? As far as I can see, openblas-linked `scipy` works well with atlas-linked `numpy`.

**Comment by Doeme** [comment]
>@Doeme: regarding BLAS_USEDEP, do we really need it? As far as I can see, openblas-linked scipy works well with atlas-linked numpy. Removing it simplifies ebuilds significantly.

Yes, It would be simpler in deed. I guess it doesn't matter that much, it just feels wrong overwriting the symbols of a library with symbols of another one, even if they are ABI and API compatible.

Anyway, even with BLAS_USEDEP we couldn't enforce 100% consistency. Consider a package not depending on BLAS but depending on two different packages that depend on BLAS. We would have to add the blas eclass to its ebuild too, just to ensure that the subgraph is consistent.

**Comment by Doeme** [comment]
So, I removed the subgraph consistency from the [GLEP](https://github.com/Doeme/NeuroGentooProject/blob/master/BLAS_Lapack/glep-%254d.rst) and updated it a bit to reflect some things discussed here (e.g. the "default to f:*")
If you don't have any bigger issues with it, I will start flooding it to the channels in the next few days, and collect feedback from the broader community.

**Comment by Doeme** [comment]
https://forums.gentoo.org/viewtopic-p-8146934.html

**Comment by mmokrejs** [comment]
@Doeme Please continue working on this. So many people are hitting the BLAS/LAPACK and also ::science vs. ::gentoo inconsistencies. Myself have a number of issues on my hosts too, jsut waiting for a clear way out. Thank you.

**Comment by heroxbd** [comment]
@Doeme, Gentoo is hosting Google summer of code 2018.  If you are interested in finalizing this blas eclass project with us by the support of Google, don't hesitate to talk to me.

**Comment by leycec** [comment]
**Dear \*NIX Gods.** The `::science` stack is blatantly superior to the Portage stack with respect to BLAS, CBLAS, and LAPACK integration. Why hasn't the former simply been merged into the latter already? Portage's scientific stack is sufficiently minimal as to be largely useless for real-world use cases.

Notably, the lack of optimized BLAS, CBLAS, and LAPACK implementations in the Portage stack constrains downstream Portage libraries (e.g., NumPy) to single- rather than multi-core usage. Our [multiphysics biology simulator](https://github.com/leycec/raiagent/tree/master/sci-biology/betse) effectively *requires* NumPy to be linked against an optimized scientific stack to ensure sanity. Since we have no means of mandating this dependency at Portage installation time, we simply emit non-fatal `pkg_pretend()` warnings redirecting end users to the [dodgy `lapack-migration.sh` script](https://github.com/gentoo/sci/blob/master/scripts/lapack-migration.sh) recently bundled by the `::science` overlay.

Clearly, this is suboptimal. End users shouldn't have to manually add unofficial overlays and run non-standard shell scripts merely to enable multicore usage in scientific applications. Multicore usage should be the well-supported default – *not* the unsupported exception. <sup>*understatement of the thread*</sup>

I'd actually iterate one step further and argue for ATLAS rather than `blas-reference`, `cblas-reference`, and `lapack-reference` as the default BLAS, CBLAS, and LAPACK implementations satisfying their respective virtuals. In late 2018, ATLAS is effectively as stable as its reference counterparts and substantially more performant. At the very least, defaulting to ATLAS would guarantee multicore-enabled efficiency under NumPy and dependents thereof (e.g., [us](https://gitlab.com/betse/betse)).

**Comment by kiwifb** [comment]
My preference is for openblas on most settings but clearly the lapack side needs improvements. But we have been talking of taking thing a step further almost a year ago without action - because mainly of real life and things.

**Comment by heroxbd** [comment]
@leycec Hi, thank you for raising this issue.  I hear you.  If you read though this issue, you would notice that we were half-way moving to a eclass mechnism, before our contributer @Doeme left and disappeared.   Right now the progress has stalled and no one has stepped up yet.

**Comment by heroxbd** [comment]
Dear @TheChymera @kiwifb @sfabbro @mmokrejs @Doeme @SoapGentoo  ,

Would you please give a review to https://archives.gentoo.org/gentoo-science/message/cbbcc2c23db1d1c14799246278d45f49 ?  This solution comes from experience of Debian contributed by @cdluminate , and is very similar to @sfabbro 's original design. In addition, it uses @mgorny 's ld.so prescription to avoid manipulating symlinks in /usr outside PM control.

It is simple enough to allow a normal user who does not care about BLAS/LAPACK uses the reference implementations by default.  Scientific users can enable the USE=eselect-ldso to unlock the full eselect switching functionality.

I regard it solves all the existing and future problems, and potentially obsoletes the framework in Gentoo main repo, the alternative-2.eclass/eselect in science overlay, and the PYTHON_TARGET-styled framework proposed by @Doeme .  I would like to see if you guys share the same judgment as mine.   Please give it a review. If you like it, please help migrating the existing ebuilds to the new framework. Mostly there will be no changes to the BLAS/LAPACK consumers, but @cdluminate is preparing a developers' guide for that.

Let's finally unify the Gentoo BLAS/LAPACK framework.  Let's terminate the long termed Gentoo main tree vs. science overlay conflict and confusion.

Cheers,
Benda

**Comment by heroxbd** [comment]
The new framework has landed on ::gentoo, please help test sci overlay packages against it.  After all the reverse dependencies are verified to work, we are going to remove alternative-2 and eselect-*-r100 from the sci overlay and unify the BLAS/LAPACK implementation.

Your help is very much needed to proceed.

Yours,
Benda

**Comment by equaeghe** [comment]
> The new framework has landed on ::gentoo, please help test sci overlay packages against it.

Starting from a stable ::gentoo BLAS/LAPACK installation, with OpenBLAS installed, what should I do to make use to the new framework? (I have already set eselect-ldso globally.)

**Comment by heroxbd** [comment]
> > The new framework has landed on ::gentoo, please help test sci overlay packages against it.
> 
> Starting from a stable ::gentoo BLAS/LAPACK installation, with OpenBLAS installed, what should I do to make use to the new framework? (I have already set eselect-ldso globally.)

Please install virtual/blas, virtual/lapack and sci-libs/lapack 3.8 versions from unstable, and follow the documentation at

  https://github.com/cdluminate/my-overlay/blob/master/usage.md

**Comment by mmokrejs** [comment]
So far I had:
```
 # emerge -pv eselect sci-libs/lapack-reference  sci-libs/cblas-reference  sci-libs/blas-reference gsl

These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild   R    ] app-admin/eselect-1.4.14::gentoo  USE="vim-syntax -doc -emacs" 0 KiB
[ebuild   R    ] sci-libs/blas-reference-20171112-r100:0/3.8.0::science  USE="-int64 -static-libs -test" ABI_X86="32 (64) (-x32)" 0 KiB
[ebuild   R    ] sci-libs/gsl-2.5-r1:0/23::gentoo  USE="deprecated -cblas-external -static-libs" 0 KiB
[ebuild   R    ] sci-libs/lapack-reference-3.8.0-r100:0/3.8.0::science  USE="deprecated -int64 -static-libs -test -xblas" ABI_X86="32 (64) (-x32)" 0 KiB
[ebuild   R    ] sci-libs/cblas-reference-20161223-r101:0/3.7.0::science  USE="-int64 -static-libs -test" ABI_X86="32 (64) (-x32)" 0 KiB

Total: 5 packages (5 reinstalls), Size of downloads: 0 KiB

...

# emerge --unmerge  sci-libs/lapack-reference sci-libs/cblas-reference sci-libs/blas-reference openblas
 * This action can remove important packages! In order to be safer, use
 * `emerge -pv --depclean <atom>` to check for reverse dependencies before
 * removing packages.

--- Couldn't find 'openblas' to unmerge.

 sci-libs/lapack-reference
    selected: 3.8.0-r100 
   protected: none 
     omitted: none 

 sci-libs/cblas-reference
    selected: 20161223-r101 
   protected: none 
     omitted: none 

 sci-libs/blas-reference
    selected: 20171112-r100 
   protected: none 
     omitted: none 

All selected packages: =sci-libs/lapack-reference-3.8.0-r100 =sci-libs/blas-reference-20171112-r100 =sci-libs/cblas-reference-20161223-r101

>>> 'Selected' packages are slated for removal.
>>> 'Protected' and 'omitted' packages will not be removed.

>>> Waiting 5 seconds before starting...
>>> (Control-C to abort)...
>>> Unmerging in: 5 4 3 2 1
>>> Unmerging (1 of 3) sci-libs/lapack-reference-3.8.0-r100...
>>> needed   sym /usr/lib64/libblas.so.3
>>> needed   obj /usr/lib64/libblas.so.3.8.0
>>> needed   sym /usr/lib64/libreflapack.so.3
>>> needed   obj /usr/lib64/libreflapack.so.3.8.0
!!! Error: Can't load module alternatives
exiting
 * eselect alternatives update returned "250"
>>> Unmerging (2 of 3) sci-libs/cblas-reference-20161223-r101...
>>> needed   obj /usr/lib64/librefcblas.so
!!! Error: Can't load module alternatives
exiting
 * eselect alternatives update returned "250"
>>> Unmerging (3 of 3) sci-libs/blas-reference-20171112-r100...
>>> needed   sym /usr/lib64/librefblas.so.3
>>> needed   obj /usr/lib64/librefblas.so.3.8.0
!!! Error: Can't load module alternatives
exiting
 * eselect alternatives update returned "250"

 * Messages for package sci-libs/lapack-reference-3.8.0-r100:

 * eselect alternatives update returned "250"

 * Messages for package sci-libs/cblas-reference-20161223-r101:

 * eselect alternatives update returned "250"

 * Messages for package sci-libs/blas-reference-20171112-r100:

 * eselect alternatives update returned "250"

 * GNU info directory index is up-to-date.
...
>>> package: sci-libs/lapack-reference-3.8.0-r100
 *  - /usr/lib64/libblas.so.3
 *  - /usr/lib64/libblas.so.3.8.0
 *      used by /usr/lib/python3.7/site-packages/numpy/linalg/_umath_linalg.cpython-37m-x86_64-linux-gnu.so (dev-python/numpy-1.16.1)
 *      used by /usr/lib/python3.7/site-packages/scipy/cluster/_vq.cpython-37m-x86_64-linux-gnu.so (sci-libs/scipy-1.1.0)
 *      used by /usr/lib/python3.7/site-packages/scipy/integrate/vode.cpython-37m-x86_64-linux-gnu.so (sci-libs/scipy-1.1.0)
 *      used by 43 other files
 *  - /usr/lib64/libreflapack.so.3
 *  - /usr/lib64/libreflapack.so.3.8.0
 *      used by /usr/lib/python3.7/site-packages/numpy/linalg/_umath_linalg.cpython-37m-x86_64-linux-gnu.so (dev-python/numpy-1.16.1)
 *      used by /usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-linux-gnu.so (dev-python/numpy-1.16.1)
 *      used by /usr/lib/python3.7/site-packages/scipy/integrate/_odepack.cpython-37m-x86_64-linux-gnu.so (sci-libs/scipy-1.1.0)
 *      used by 68 other files
...
>>> package: sci-libs/cblas-reference-20161223-r101
 *  - /usr/lib64/librefcblas.so
 *      used by /usr/lib/python3.7/site-packages/numpy/core/_multiarray_umath.cpython-37m-x86_64-linux-gnu.so (dev-python/numpy-1.16.1)
 *      used by /usr/lib64/python2.7/site-packages/numpy/core/_multiarray_umath.so (dev-python/numpy-1.16.1)
 *      used by /usr/lib64/python3.6/site-packages/numpy/core/_multiarray_umath.cpython-36m-x86_64-linux-gnu.so (dev-python/numpy-1.16.1)

```

So I am going to test:

```
# emerge -pv blas lapack

These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild  N     ] sci-libs/lapack-3.8.0::gentoo  USE="-doc -eselect-ldso -lapacke" 0 KiB
[ebuild     U  ] virtual/lapack-3.8::gentoo [3.7-r100::science] USE="-eselect-ldso% (-doc%) (-int64%)" ABI_X86="(-32%*) (-64%*) (-x32%)" 0 KiB
[ebuild     U  ] virtual/blas-3.8::gentoo [3.7-r100::science] USE="-eselect-ldso% (-doc%) (-int64%)" ABI_X86="(-32%*) (-64%*) (-x32%)" 0 KiB
```

```
...
>>> Installing (1 of 3) sci-libs/lapack-3.8.0::gentoo
...
 * 
 * Detected file collision(s):
 * 
 * 	/usr/include/cblas.h
 * 	/usr/lib64/pkgconfig/cblas.pc
 * 	/usr/lib64/pkgconfig/lapack.pc
 * 	/usr/lib64/liblapack.so
 * 	/usr/lib64/libcblas.so
 * 
 * Searching all installed packages for file collisions...
 * 
 * Press Ctrl-C to Stop
 * 
 * None of the installed packages claim the file(s).
 * 
 * Package 'sci-libs/lapack-3.8.0' merged despite file collisions. If
 * necessary, refer to your elog messages for the whole content of the
 * above message.

>>> Emerging (2 of 3) virtual/lapack-3.8::gentoo
>>> Unpacking source...
>>> Source unpacked in /scratch/var/tmp/portage/virtual/lapack-3.8/work
>>> Preparing source in /scratch/var/tmp/portage/virtual/lapack-3.8/work ...
>>> Source prepared.
>>> Configuring source in /scratch/var/tmp/portage/virtual/lapack-3.8/work ...
>>> Source configured.
>>> Compiling source in /scratch/var/tmp/portage/virtual/lapack-3.8/work ...
>>> Source compiled.
>>> Test phase [not enabled]: virtual/lapack-3.8

>>> Install virtual/lapack-3.8 into /scratch/var/tmp/portage/virtual/lapack-3.8/image
>>> Completed installing virtual/lapack-3.8 into /scratch/var/tmp/portage/virtual/lapack-3.8/image

 * Final size of build directory: 4 KiB
 * Final size of installed tree:  4 KiB


>>> Installing (2 of 3) virtual/lapack-3.8::gentoo
...
```

Done, but now:

```
# emerge @preserved-rebuild --keep-going

 * IMPORTANT: 4 news items need reading for repository 'gentoo'.
 * Use eselect news read to view new items.

Calculating dependencies... done!
[ebuild   R    ] sci-libs/arpack-3.5.0 
[ebuild   R    ] sci-libs/cholmod-2.1.2 
[ebuild  N     ] sci-libs/cblas-reference-20161223-r101  USE="-int64 -static-libs -test" ABI_X86="32 (64) (-x32)" 
[ebuild   R    ] sci-libs/coinor-utils-2.9.11 
[ebuild   R    ] sci-geosciences/mapnik-3.0.18 
[ebuild   R    ] www-servers/apache-2.2.34 
[ebuild   R    ] dev-lang/R-3.6.0 
[ebuild   R    ] sci-libs/spqr-1.3.1 
[ebuild   R    ] dev-python/numpy-1.16.1 
[ebuild   R    ] sci-libs/scipy-1.1.0 
[blocks B      ] sci-libs/cblas-reference ("sci-libs/cblas-reference" is blocking sci-libs/lapack-3.8.0)

 * Error: The above package list contains packages which cannot be
 * installed at the same time on the same system.

  (sci-libs/lapack-3.8.0:0/0::gentoo, installed) pulled in by
    >=sci-libs/lapack-3.8.0[eselect-ldso?] (>=sci-libs/lapack-3.8.0) required by (virtual/lapack-3.8:0/0::gentoo, installed)
    >=sci-libs/lapack-3.8.0[-eselect-ldso] required by (virtual/blas-3.8:0/0::gentoo, installed)

  (sci-libs/cblas-reference-20161223-r101:0/3.7.0::science, ebuild scheduled for merge) pulled in by
    >=sci-libs/cblas-reference-20161223-r100[int64?,abi_x86_32(-)?,abi_x86_64(-)?,abi_x86_x32(-)?,abi_mips_n32(-)?,abi_mips_n64(-)?,abi_mips_o32(-)?,abi_riscv_lp64d(-)?,abi_riscv_lp64(-)?,abi_s390_32(-)?,abi_s390_64(-)?] (>=sci-libs/cblas-reference-20161223-r100[abi_x86_32(-),abi_x86_64(-)]) required by (virtual/cblas-3.7-r100:0/0::science, installed)


```

I uninstalled `virtual/cblas-3.7-r100` and the rest went smoothly. I did not hit any problems during the transition. Thank you for your efforts, @heroxbd .

**Comment by lkraav** [comment]
Curious, how can https://github.com/gentoo/sci/blob/master/sci-misc/kaldi/kaldi-5.1.65.ebuild work with the updated platform now?

```
>>> Running pre-merge checks for sci-misc/kaldi-5.1.65
!!! Error: Can't load module cblas
exiting
 * ERROR: sci-misc/kaldi-5.1.65::science failed (pretend phase):
 *   Build with '' CBLAS is not supported
 * 
 * Call stack:
 *             ebuild.sh, line 124:  Called pkg_pretend
 *   kaldi-5.1.65.ebuild, line  45:  Called die
 * The specific snippet of code:
 *              die "Build with '${cblas_provider}' CBLAS is not supported"
 * 
 * If you need support, post the output of `emerge --info '=sci-misc/kaldi-5.1.65::science'`,
 * the complete build log and the output of `emerge -pqv '=sci-misc/kaldi-5.1.65::science'`.
 * The complete build log is located at '/mnt/datapool/tmp/portage/sci-misc/kaldi-5.1.65/temp/build.log'.
 * The ebuild environment file is located at '/mnt/datapool/tmp/portage/sci-misc/kaldi-5.1.65/temp/die.env'.
 * Working directory: '/mnt/datapool/tmp/portage/sci-misc/kaldi-5.1.65/homedir'
 * S: '/mnt/datapool/tmp/portage/sci-misc/kaldi-5.1.65/work/kaldi-9e06caa1fb0306898632c6fa3ad67571c4d06cf5/src'
```

**Comment by kiwifb** [comment]
The reliance on either `openblas` or `atlas` for `kaldi` has always been deplorable. I guess it is an opportunity to just fix it.

**Comment by heroxbd** [comment]
> I uninstalled `virtual/cblas-3.7-r100` and the rest went smoothly. I did not hit any problems during the transition. Thank you for your efforts, @heroxbd .

Let's thank @cdluminate, he did all the great jobs.

**Comment by kiwifb** [comment]
I just migrated my system today and sage-on-gentoo works fine with it. And that implies a good number of packages. It was definitely something. 

Great job.

**Comment by leycec** [comment]
**Phenomenal work,** @cdluminate, @heroxbd, and friends. I can only parrot prior comments. The upgrade process is surprisingly seamless, the [documentation is surprisingly human-readable](https://github.com/cdluminate/my-overlay/blob/master/usage.md), and `eselect-ldso` appears to behave exactly as expected. In short, scientific Shangri-La is now is sight.

If `eselect-ldso` continues to behave exactly as expected for all involved, are there any plans to enable this functionality by default at some future date (e.g., either by enabling `+eselect-ldso` and `eselect-ldso(+)` across all ebuilds *or* simply dropping `eselect-ldso` and forcing this behaviour)? Unless I'm shamefully mistaken, <sup>*it happens alot*</sup> `eselect-ldso` has significant advantages with no corresponding disadvantages by compare to the default `-eselect-ldso` approach – which is nice.

Thanks again, all! You are true computational heroes. :full_moon_with_face:

**Comment by heroxbd** [comment]
> **Phenomenal work,** @cdluminate, @heroxbd, and friends. I can only parrot prior comments. The upgrade process is surprisingly seamless, the [documentation is surprisingly human-readable](https://github.com/cdluminate/my-overlay/blob/master/usage.md), and `eselect-ldso` appears to behave exactly as expected. In short, scientific Shangri-La is now is sight.

Awesome!  There is nothing more rewarding than happy users for us!

> If `eselect-ldso` continues to behave exactly as expected for all involved, are there any plans to enable this functionality by default at some future date (e.g., either by enabling `+eselect-ldso` and `eselect-ldso(+)` across all ebuilds _or_ simply dropping `eselect-ldso` and forcing this behaviour)? Unless I'm shamefully mistaken, _it happens alot_ `eselect-ldso` has significant advantages with no corresponding disadvantages by compare to the default `-eselect-ldso` approach – which is nice.

It is possible to enable +eselect-ldso by default on common profiles.  But for Gentoo targets without glibc, like FreeBSD, MacOS, and Prefix/rpath, the USE flag is not tested or not implemented.

Let's keep the present settings for a while and see how it goes.

> Thanks again, all! You are true computational heroes. full_moon_with_face

The pleasure is mine.

Benda

**Comment by waebbl** [comment]
Solid work all. The process of switching is straight-forward and well documented and works pretty well.

I have three packages on my system, which probably need updating, as they still want to pull in `virtual/cblas`: `dev-python/numpy[lapack]`, `sci-libs/scipy` and `sci-libs/scikits_learn`.

Thanks for the good work!

**Comment by cdluminate** [comment]
I copied the `usage.md` to wiki.gentoo.org: https://wiki.gentoo.org/wiki/Blas-lapack-switch
